FEXHeaderUtils/FEXHeaderUtils/Filesystem.h

```diff
@@ -14,4 +22,270 @@ namespace FHU::Filesystem {
+  }
+
+  inline bool IsRelative(const fextl::string &Path) {
+    return Path.starts_with('/');
+  }
+
+  enum class CopyOptions {
+    NONE,
+    SKIP_EXISTING,
+    OVERWRITE_EXISTING,
+  };
+
+  inline bool CopyFile(const fextl::string &From, const fextl::string &To, CopyOptions Options = CopyOptions::NONE) {
+    bool DestTested = false;
+    bool DestExists = false;
+    if (Options == CopyOptions::SKIP_EXISTING && Exists(To.c_str())) {
+      DestTested = true;
+      DestExists = Exists(To.c_str());
+      if (DestExists) {
+        // If the destination file exists already and the skip existing flag is set then
+        // return true without error.
+        return true;
+      }
+    }
+
+    if (!DestTested) {
+      DestTested = true;
+      DestExists = Exists(To.c_str());
+    }
+
+    if (Options == CopyOptions::OVERWRITE_EXISTING && DestExists) {
+      // If we are overwriting and the file exists then we want to use `sendfile` to overwrite
+      int SourceFD = open(From.c_str(), O_RDONLY | O_CLOEXEC);
+      if (SourceFD == -1) {
+        return false;
+      }
+
+      int DestinationFD = open(To.c_str(), O_WRONLY | O_CREAT | O_TRUNC, 0200);
+      if (DestinationFD == -1) {
+        close(SourceFD);
+        return false;
+      }
+
+      struct stat buf;
+      if (fstat(SourceFD, &buf) != 0) {
+        close(DestinationFD);
+        close(SourceFD);
+        return false;
+      }
+
+      // Set the destination permissions to the original source permissions.
+      if (fchmod(DestinationFD, buf.st_mode) != 0) {
+        close(DestinationFD);
+        close(SourceFD);
+        return false;
+      }
+      bool Result = sendfile(DestinationFD, SourceFD, nullptr, buf.st_size) == buf.st_size;
+      close(DestinationFD);
+      close(SourceFD);
+      return Result;
+    }
+
+    if (!DestExists) {
+      // If the destination doesn't exist then just use rename.
+      return rename(From.c_str(), To.c_str()) == 0;
+    }
+
+    return false;
+  }
+
+  inline fextl::string LexicallyNormal(const fextl::string &Path) {
+    const auto PathSize = Path.size();
+
+    // Early exit on empty paths.
+    if (PathSize == 0) {
+      return {};
+    }
+
+    // Count the number of separators up front
+    const auto SeparatorCount = std::count(Path.begin(), Path.end(), '/');
+
+    // Needs to be a list so iterators aren't invalidated while erasing.
+    // Maximum number of list regions will be the counted separators plus 2.
+    // Multiplied by two since it allocates * 2 the size (32-bytes per element)
+    // Use a small stack allocator to be more optimal.
+    // Needs to be a list so iterators aren't invalidated while erasing.
+    size_t DataSize = sizeof(std::string_view) * (SeparatorCount + 2) * 2;
+    void *Data = alloca(DataSize);
+    fextl::pmr::fixed_size_monotonic_buffer_resource mbr(Data, DataSize);
+    std::pmr::polymorphic_allocator pa {&mbr};
+    std::pmr::list<std::string_view> Parts{pa};
```

Comment:
Instead of doing this dance to allocate a list on the stack, couldn't a `fextl::vector` be used with indexes instead?

Reply:
We're walking iterators and deleting iterators as we're walking the list. Which doesn't mesh what vector gives us.

Reply:
Also this is specifically done on the stack for performance reasons. This function is absolutely hammered by pressure-vessel and using this custom pmr allocator. Microbench showing that a regular list without pmr is 2.6x slower

Reply:
> We're walking iterators and deleting iterators as we're walking the list. Which doesn't mesh what vector gives us.

Does "doesn't mesh well" mean it literally won't work or you anticipate performance to be worse? We're not *storing* more than one iterator at a time though, so I think it should be fine as long as we use indexes rather than iterators.

> Also this is specifically done on the stack for performance reasons. This function is absolutely hammered by pressure-vessel and using this custom pmr allocator is worth it. Microbench showing that a regular list without pmr is 2.6x slower

Sounds fine. If you have spare time I'm curious how vector would perform with the same stack-allocation trick. There'd be a bunch of small memory reallocations, but we'd save on pointer traversals for the majority of iteration steps that don't insert/erase. (and we also wouldn't have std::list's internal representation leak into our implementation through the size estimate.)

Reply:
I don't see how it would work with the idea of iterator deletion without shuffling the entire vector each time an erase occurs.

If the issue of leaking std::list implementation details is the concern then wiring up a doubly linked list that is guaranteed to be 2x pointers + object size by providing a static element size query would resolve this?

Reply:
> I don't see how it would work with the idea of iterator deletion without shuffling the entire vector each time an erase occurs.

You would have to relocate all memory past the erased element, yes. (std::vector::erase does this for you)

This is fairly lightweight since we're only relocating `string_views`, and more importantly it's relatively rare (unless there are more `.`/`..` nodes in the path than actual directories). So I'd still expect vector to perform better on average than list here, since the latter will do pointer lookups crossing the L1 cacheline size every second iteration.

Reply:
I really don't want to implement a new container paradigm for a task that is already a perfect fit for a doubly linked list with a fixed maximum number of nodes. If you're itching to implement new container paradigms for microbenching then go for it.
The farthest I would go is implementing a `fixed_list` type to avoid leaking `std::list` implementation details.